src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt

```diff
@@ -0,0 +1,259 @@
+                    TextItem(it, mimeType)
+                }
+            }
+    }
+
+    private fun constructResource(title: Content, body: Content?): Resource? {
+        val titleTextItem = textItem(title)
+            ?: return null
+
+        return Resource(
+            sort = title.sort,
+            title = titleTextItem,
+            body = textItem(body),
+            titleAudio = constructAssociatedAudio(title),
+            bodyAudio = body?.let { constructAssociatedAudio(body) }
+        )
+    }
+
+    private fun constructResourceGroups(content: Content): List<ResourceGroup> {
+        return constructResourceGroups(
+            rcInfoList = resourceRepository.getResourceContainerInfo(content),
+            getResourceContents = { resourceRepository.getResources(content, it) }
+        )
+    }
+
+    private fun constructResourceGroups(collection: Collection): List<ResourceGroup> {
+        return constructResourceGroups(
+            rcInfoList = resourceRepository.getResourceContainerInfo(collection),
+            getResourceContents = { resourceRepository.getResources(collection, it) }
+        )
+    }
+
+    private fun constructResourceGroups(
+        rcInfoList: List<ResourceContainerInfo>,
+        getResourceContents: (ResourceContainerInfo) -> Observable<Content>
+    ): List<ResourceGroup> {
+        return rcInfoList.map {
+            ResourceGroup(
+                it,
+                getResourceContents(it)
+                    .contentsToResources()
+                    .cache()
+            )
+        }
+    }
+
+    private fun Observable<Content>.contentsToResources(): Observable<Resource> {
+        return this
+            .buffer(2, 1)
+            .concatMapMaybe { (a, b) ->
+                Maybe.fromCallable {
+                    when {
+                        a.labelKey != "title" -> null
+                        b.labelKey != "body" -> constructResource(a, null)
+                        else -> constructResource(a, b)
+                    }
+                }
+            }
+    }
+
+    /** Build a relay primed with the current deletion state, that responds to updates by writing to the DB. */
+    private fun deletionRelay(modelTake: ModelTake): BehaviorRelay<DateHolder> {
+        val relay = BehaviorRelay.createDefault(DateHolder(modelTake.deleted))
+
+        val subscription = relay
+            .skip(1) // ignore the initial value
+            .subscribe {
+                takeRepository.update(modelTake.copy(deleted = it.value))
+            }
+
+        connections += subscription
+        return relay
+    }
+
+    private fun deselectUponDelete(take: WorkbookTake, selectedTakeRelay: BehaviorRelay<TakeHolder>) {
+        val subscription = take.deletedTimestamp
+            .filter { localDate -> localDate.value != null }
+            .filter { take == selectedTakeRelay.value?.value }
+            .map { TakeHolder(null) }
+            .subscribe(selectedTakeRelay)
```

Comment:
IMO, the concept behind `deselectUponDelete` is brilliant, but difficult for me personally to understand without comments. Though I think it is difficult simply because I dislike the way that `subscribe(consumer)` LOOKS syntactically. The way I think of it is that selectedTakeRelay is subscribing TO deletedTimestamp, but this makes it look like deletedTimestamp is subscribing to selectedTakeRelay. I would have named the `subscribe` method something more unambiguous, but perhaps it just takes a little getting used to!

Reply:
I agree.  `a.subscribe(b)` should mean `a` wants to be notified of `b`'s events.  That's not how Rx uses it though.